llmfoundry/data/packing.py

```diff
@@ -260,124 +261,95 @@ def pad_tensor(tensor: torch.Tensor, pad_value: int):
-        parser.add_argument(
-            '--num-packing-ratios',
-            type=int,
-            default=10,
-            help=
-            'Number of packing_ratio values (spaced between `min` and `max) to try.'
-        )
-
-        args = parser.parse_args()
-
-        if not os.path.isfile(args.yaml_path):
-            raise FileNotFoundError(
-                '`yaml_path` does not correspond to any existing file.')
-        if args.num_devices < 1:
-            raise ValueError('`num_devices` must be a positive integer.')
-        if args.min < 1.0:
-            raise ValueError('`min` must be >=1.0.')
-        if args.max < args.min:
-            raise ValueError('`max` cannot be less than `min`.')
-        if args.num_packing_ratios < 1:
-            raise ValueError('`num_packing_ratios` must be a positive integer.')
-        return args
-
-    def build_dataloader(cfg: DictConfig, tokenizer: PreTrainedTokenizerBase,
-                         device_batch_size: int):
-        if cfg.name == 'text':
-            return build_text_dataloader(cfg, tokenizer, device_batch_size)
-        elif cfg.name == 'text_denoising':
-            return build_text_denoising_dataloader(cfg, tokenizer,
-                                                   device_batch_size)
-        elif cfg.name == 'finetuning':
-            return build_finetuning_dataloader(cfg, tokenizer,
-                                               device_batch_size)
-        else:
-            raise ValueError(
-                f'Not sure how to build dataloader with config: {cfg}')
-
-    args = parse_args()
+def auto_packing_ratio(dataloader_cfg: DictConfig,
+                       tokenizer: PreTrainedTokenizerBase,
+                       device_batch_size: int):
+    """Find a packing ratio that minimizes padding with zero waste.
+
+    Args:
+        dataloader_cfg (DictConfig): The dataloader configuration for profiling.
+        tokenizer (PreTrainedTokenizerBase): The tokenizer for profiling.
+        device_batch_size (int): The size of the batches (number of examples) per device.
+
+    Returns:
+        A packing ratio that minimizes padding while maintaining zero waste.
+    """
+    from composer.utils import dist, get_device
+    min_ratio = 1
+    max_ratio = dataloader_cfg.dataset.max_seq_len / 100
+    num_packing_ratios = 20
+    profiling_results = profile_packing(dataloader_cfg, tokenizer, min_ratio,
+                                        max_ratio, num_packing_ratios,
+                                        device_batch_size)
+
+    # Obtain the maximum packing_ratio/minimum padding that has no waste.
+    packing_ratio = 1
+    for packing_ratio_candidate, _, waste in profiling_results:
+        if waste > 0:
+            break
+        packing_ratio = packing_ratio_candidate
+
+    # Select the minimum packing ratio across all ranks.
+    if torch.cuda.is_available() and dist.is_available(
```

Comment:
I think the way to do this generically is

```
if dist.is_available() and dist.is_initialized():
    device = get_device(None)
    packing_ratio_tensor = device.tensor_to_device(torch.tensor(packing_ratio))
    dist.all_reduce(packing_ratio_tensor, reduce_operation='MIN')
    packing_ratio = packing_ratio_tensor.item()
```

Reply:
when I use dist without cuda I get errors--is it supposed to work with other backends? (I think yes, but I haven't gotten it to work...)

Reply:
yes, it should work with gloo backend as well. but this is not a big deal to investigate. can you make a jira to fix this and put it in the llm foundry backlog epic?

Reply:
actually nvm, i'm not sure what i was doing before, but i tested it out again with gloo just now and it works fine. I suspect I was running nccl x cpu previously. 

```
from composer.utils import dist, get_device
import torch

torch.distributed.init_process_group('gloo')

if dist.is_available() and dist.is_initialized(): 
    device = get_device(None) 
    if dist.get_local_rank() == 0:
        packing_ratio_tensor = device.tensor_to_device(torch.tensor(5.0)) 
    else:
        packing_ratio_tensor = device.tensor_to_device(torch.tensor(4.0)) 

    dist.all_reduce(packing_ratio_tensor, reduce_operation='MIN') 
    packing_ratio = packing_ratio_tensor.item()
    print(packing_ratio)
```

Reply:
will update to generic